_source/_posts/2022-10-01-spring-vault.md

```diff
@@ -0,0 +1,639 @@
+
+```shell
+jshell
+
+UUID.randomUUID()
+```
+
+```shell
+ENCRYPT_KEY={encryptKey} ./mvnw spring-boot:run
+```
+
+Then, in another terminal, encrypt your client ID and secret.
+
+```shell
+http :8888/encrypt --raw {yourOktaClientId}
+http :8888/encrypt --raw {yourOktaClientSecret}
+```
+
+In the `vault-config-server` project folder, create a `src/main/resources/config/vault-demo-app-dev.yml` file to store the secrets for the `dev` profile, with the following contents:
+
+```yml
+okta:
+  oauth2:
+    issuer: {yourIssuerURI}
+    clientId: '{cipher}encryptedClientId'
+    clientSecret: '{cipher}encryptedClientSecret'
+```
+
+The `client-id` and `client-secret` encrypted values must be prefixed with `{cipher}`. Restart the config server.
+
+To consume the config server properties, the client application must set the server address in the application configuration. In the `vault-demo-app` project, rename `application.properties` to `application.yml` and configure it as follows:
+
+```yaml
+spring:
+  cloud:
+    config:
+      uri: http://localhost:8888
+  config:
+    import: "configserver:"
+  application:
+    name: vault-demo-app
+  profiles:
+    active: dev
+```
+
+Start `vault-demo-app` without passing the environment variables:
+
+```shell
+./mvnw clean
+./mvnw spring-boot:run
+```
+
+When requesting `http://localhost:8080` it should again redirect to the Okta login.
+
+In a real environment, the config server should be secured. Spring Cloud Config Server supports asymmetric key encryption as well, with the server encrypting with the public key, and the clients decrypting with the private key. However, the documentation warns about spreading the key management process around clients.
+
+## Vault as a Configuration Backend with Spring Cloud Vault
+
+{% img blog/spring-vault/vault-logo.png alt:"Vault logo" width:"300" %}{: .pull-right }
+
+In the cloud, secrets management has become much more difficult. Vault is a secrets management and data protection tool from HashiCorp that provides secure storage, dynamic secret generation, data encryption, and secret revocation.
+
+Vault encrypts the secrets prior to writing them to persistent storage. The encryption key is also stored in Vault, but encrypted with a _master key_ not stored anywhere. The master key is split into shards using _Shamir's Secret Sharing algorithm_, and distributed among a number of operators. The Vault unseal process allows you to reconstruct the master key by adding shards one at a time in any order until enough shards are present, then Vault becomes operative. Operations on secrets can be audited by enabling audit devices, which will send audit logs to a file, syslog or socket.
```

Comment:
Should "Shamir’s Secret Sharing algorithm" be linked to?

Reply:
Added a link.